homeassistant/components/bluetooth/models.py

```diff
@@ -0,0 +1,138 @@
+"""Models for bluetooth."""
+from __future__ import annotations
+
+import asyncio
+import contextlib
+import logging
+from typing import Any, Final, cast
+
+from bleak import BleakScanner
+from bleak.backends.device import BLEDevice
+from bleak.backends.scanner import AdvertisementData, AdvertisementDataCallback
+from lru import LRU  # pylint: disable=no-name-in-module
+
+from homeassistant.core import CALLBACK_TYPE, callback as hass_callback
+
+_LOGGER = logging.getLogger(__name__)
+
+FILTER_UUIDS: Final = "UUIDs"
+
+HA_BLEAK_SCANNER: HaBleakScanner | None = None
+
+MAX_HISTORY_SIZE: Final = 512
+
+
+def _dispatch_callback(
+    callback: AdvertisementDataCallback,
+    filters: dict[str, set[str]],
+    device: BLEDevice,
+    advertisement_data: AdvertisementData,
+) -> None:
+    """Dispatch the callback."""
+    if not callback:
+        # Callback destroyed right before being called, ignore
+        return
+
+    if (uuids := filters.get(FILTER_UUIDS)) and not uuids.intersection(
+        advertisement_data.service_uuids
+    ):
+        return
+
+    try:
+        callback(device, advertisement_data)
+    except Exception:  # pylint: disable=broad-except
+        _LOGGER.exception("Error in callback: %s", callback)
+
+
+class HaBleakScanner(BleakScanner):  # type: ignore[misc]
+    """BleakScanner that cannot be stopped."""
+
+    # HaBleakScanner is a singleton
+    _callbacks: list[tuple[AdvertisementDataCallback, dict[str, set[str]]]] = []
+    _history: LRU = LRU(MAX_HISTORY_SIZE)
```

Comment:
Even if it's a singleton, wouldn't it be safer for tests and stuff to just instantiate this in the constructor?

Reply:
Oh I guess we're extending Bleak so it now needs to wrap the constructor

Reply:
6aaa30d9036936e306eb9fbf7fcc182e82d69497